tests/test_builders/xpath_util.py

```diff
@@ -0,0 +1,56 @@
+from __future__ import annotations
+
+import re
+from typing import TYPE_CHECKING
+from xml.etree.ElementTree import tostring
+
+if TYPE_CHECKING:
+    import os
+    from collections.abc import Callable, Sequence
+    from xml.etree.ElementTree import Element
+
+
+def _gettext(node: Element) -> str:
+    if node.text is not None:
+        # the node has only one text
+        return node.text
+
+    # the node has tags and text; gather texts just under the node
+    return ''.join(n.tail or '' for n in node)
+
+
+def check_xpath(
+    root: Element,
+    filename: str | os.PathLike[str],
+    xpath: str,
+    check: str | re.Pattern[str] | Callable[[Sequence[Element]], None] | None,
+    be_found: bool = True,
+) -> None:
+    nodes = root.findall(xpath)
+    assert isinstance(nodes, list)
+
+    if check is None:
+        # check for an empty list
+        assert nodes == [], f'found any nodes matching xpath {xpath!r} in file {filename}'
+        return
+
+    assert nodes, 'did not find any node matching xpath {xpath!r} in file {filename}'
+    if not check:
+        # only check for node presence
+        return
+
+    if callable(check):
+        check(nodes)
+        return
+
+    rex = re.compile(check)
+    if be_found:
+        if any(rex.search(_gettext(node)) for node in nodes):
+            return
+    else:
+        if all(not rex.search(_gettext(node)) for node in nodes):
+            return
+
+    ctx: list[bytes] = [tostring(node, encoding='utf-8') for node in nodes]
```

Comment:
```suggestion
    ctx: list[str] = [tostring(node, encoding='unicode', method='html') for node in nodes]
```

Reply:
I'm aware of problems during parsing of X(HT)ML documents, but hadn't really thought much about problems serializing/outputting them.  The in-memory representation should be fairly static, and describes the entire node graph -- there aren't circular paths or reroutings to follow..

..even so, it's fair to be cautious.

For some reason I'm not keen on printing raw `bytes` to output, though (even if that occurs indirectly via an `Exception` type).

Reply:
One line per node would be preferable, I think (most of the selected node-lists should be short.. and maybe we should stop when the output exceeds [`80*8` chars](https://docs.python.org/3/library/unittest.html#unittest.TestCase.maxDiff)?).

Reply:
@jayaddison I've edited the answer.

Reply:
And what do you mean by "one node" per line? because sometimes the node matched by `xpath` might be VERY VERY long

Reply:
Ok, yep, fair point.  I'll try setting `check_xpath` to fail every time and print `[:640]` (not the 'production' way to do it.. better to early-exit when the limit is exceeded) from each node on a separate line in a few minutes and then running our entire test suite - I expect it'll produce fairly readable output, but I may be wrong.  I'll confirm in a little while.

Reply:
With the simplest formatter:

```
E       AssertionError: 'Baz o' not found in any node matching './/h4' in file index.html:
E       [i=0] <h4><span class="section-number">1.1.1. </span>Foo A1<a class="headerlink" href="#foo-a1" title="Link to this heading">¶</a></h4>
E       [i=1] <h4><span class="section-number">1.2.1. </span>Foo B1<a class="headerlink" href="#foo-b1" title="Link to this heading">¶</a></h4>
E       [i=2] <h4><span class="section-number">2.1.1. </span>Baz A<a class="headerlink" href="#baz-a" title="Link to this heading">¶</a></h4>
E       [i=3] <h4><span class="section-number">2.2.1. </span>Bar B1<a class="headerlink" href="#bar-b1" title="Link to this heading">¶</a></h4>
```

Maybe... it's better to have a single line. I don't need to add complex logic. And otherwise people can debug directly by looking at the sources... Yes I'll keep it simple.

Reply:
What I've got locally:

```
E       AssertionError: '^fig1$' not found in any node matching './/li/p/code/sp
an' in file index.html: [
E         <span class="pre">fig1</span>
E         <span class="pre">Figure%s</span>
E         <span class="pre">table-1</span>
E         <span class="pre">Table:%s</span>
E         <span class="pre">CODE_1</span>
E         <span class="pre">Code-%s</span>
E         <span class="pre">index</span>
E         <span class="pre">invalid</span>
E         <span class="pre">Fig</span> 
E         <span class="pre">%s</span> 
E         <span class="pre">%s</span>
E         <span class="pre">Fig.{number}</span> 
E         <span class="pre">{name}</span>
E       ]
```

Reply:
(bad because it looks like a Python `repr` output.. but isn't valid Python; are there better symbols than `[` and `]` to place around XML, I wonder?)

Reply:
I've used a different format and cut off the output at 640 characters (for each printed node)